### PR TITLE
slice_update op

### PR DESCRIPTION
**Pull request by leaves-zwx**
实现形如 
```
x = np.random.rand(2, 10)
a = np.random.rand(2, 5)
x[:, 2:7] = a
```
的功能

### CODE REVIEW

File: oneflow/user/ops/slice_op.cpp [hunk 1]
```diff
@@ -156,6 +156,84 @@ void InferSliceGradInputArgModifier(user_op::GetInputArgModifier GetInputArgModi
...
+  ctx->NewBuilder()
+      .PartialSum(user_op::OpArg("x", 0))
+      .Broadcast(user_op::OpArg("update", 0))
+      .PartialSum(user_op::OpArg("y", 0))
+      .Build();
+  ctx->NewBuilder()
+      .Broadcast(user_op::OpArg("x", 0))
+      .PartialSum(user_op::OpArg("update", 0))
+      .Broadcast(user_op::OpArg("y", 0))
+      .Build();
```
Comment (guo-ran) [inline oneflow/user/ops/slice_op.cpp:233]: 这个好像也不支持吧？

**Comment by JiaqingFu** [comment]
嗨，这个pr什么时候能合并进去哈，这边要用到这个

File: oneflow/user/ops/slice_op.cpp [hunk 2]
```diff
@@ -156,6 +156,84 @@ void InferSliceGradInputArgModifier(user_op::GetInputArgModifier GetInputArgModi
   like_modifier->set_requires_grad(false);
 }
 
+Maybe<void> InferSliceUpdateOpTensorDesc(user_op::InferContext* ctx) {
+  const Shape* x_shape = ctx->Shape4ArgNameAndIndex("x", 0);
+  DataType x_dtype = *ctx->Dtype4ArgNameAndIndex("x", 0);
+  const int64_t ndim = x_shape->NumAxes();
+  Shape* y_shape = ctx->Shape4ArgNameAndIndex("y", 0);
+  *y_shape = *x_shape;
+  *ctx->Dtype4ArgNameAndIndex("y", 0) = x_dtype;
```
[the remark this answers is not in the document]
Comment (leaves-zwx) [reply oneflow/user/ops/slice_op.cpp]: 已修改

File: oneflow/user/ops/slice_op.cpp [hunk 3]
```diff
@@ -156,6 +156,84 @@ void InferSliceGradInputArgModifier(user_op::GetInputArgModifier GetInputArgModi
...
+    }
+    const int64_t diff = (step > 0) ? (stop - start - 1) : (stop - start + 1);
+    const int64_t sliced_dim_size = diff / step + 1;
+    CHECK_EQ_OR_RETURN(sliced_dim_size, update_shape->At(i))
+        << "sliced dim size " << sliced_dim_size << " at axis " << i
+        << " not equal to the update shape " << update_shape->ToString();
+  }
+  // the split axis can't be sliced
+  const SbpParallel& x_sbp = ctx->SbpParallel4ArgNameAndIndex("x", 0);
+  if (ctx->parallel_ctx().parallel_num() != 1 && x_sbp.has_split_parallel()) {
```
[the remark this answers is not in the document]
Comment (leaves-zwx) [reply oneflow/user/ops/slice_op.cpp:194]: 在 split axis 上只能支持 full slice，这里的判断还是有必要的，因为我们通常都是 S(0) 并行，这种情况下，用户 slice dim0 就会出错，这种情况比较常见，所以直接在这里 check，提前把错误暴露。

File: oneflow/user/ops/slice_op.cpp [hunk 4]
```diff
@@ -156,6 +156,84 @@ void InferSliceGradInputArgModifier(user_op::GetInputArgModifier GetInputArgModi
...
+    if (IsFullSlice(start_vec.at(i), stop_vec.at(i), step_vec.at(i), x_shape.At(i))) {
+      ctx->NewBuilder().Split(ctx->inputs(), i).Split(ctx->outputs(), i).Build();
+    }
+  }
+  ctx->NewBuilder().PartialSum(ctx->inputs()).PartialSum(ctx->outputs()).Build();
+  ctx->NewBuilder()
+      .PartialSum(user_op::OpArg("x", 0))
+      .Broadcast(user_op::OpArg("update", 0))
+      .PartialSum(user_op::OpArg("y", 0))
+      .Build();
```
[the remark this answers is not in the document]
Comment (leaves-zwx) [reply oneflow/user/ops/slice_op.cpp]: 是的。

File: oneflow/user/ops/slice_op.cpp (hunk 1 above)
Comment (leaves-zwx) [reply oneflow/user/ops/slice_op.cpp:227]: 这个好像支持。考虑如下例子：
x: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
slice [2:5]

update 假设有 2 个 part 分别是 [10, 11, 0] 和 [0, 0, 12]

最后的结果仍然是
y: [0, 1, 10, 11, 12, 5, 6, 7, 8, 9]

Comment (guo-ran) [reply oneflow/user/ops/slice_op.cpp:227]: y1 + y2 的话除了update的其他部分也要加，结果不对了。

Comment (leaves-zwx) [reply oneflow/user/ops/slice_op.cpp:227]: 哦哦，是的。

**Review by guo-ran — APPROVED** [review]
